File: go/chromahash_test.go
package chromahash

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"os"
	"testing"
)

// The Go package is now a thin cgo wrapper over the chromahash-c C ABI, so these
// tests exercise the public API end-to-end against the shared spec vectors —
// the cross-language parity gate. Per-function unit tests of the algorithm live
// in the Rust core (the single source of truth).

// ── helpers (shared with batch_test.go / batch_bench_test.go) ─────────────────

func solidImage(w, h int, r, g, b, a byte) []byte {
	rgba := make([]byte, w*h*4)
	for i := 0; i < w*h; i++ {
		rgba[i*4] = r
		rgba[i*4+1] = g
		rgba[i*4+2] = b
		rgba[i*4+3] = a
	}
	return rgba
}

func horizontalGradient(w, h int) []byte {
	rgba := make([]byte, w*h*4)
	for y := 0; y < h; y++ {
		for x := 0; x < w; x++ {
			t := float64(x) / math.Max(float64(w-1), 1)
			idx := (y*w + x) * 4
			rgba[idx] = byte(t * 255)
			rgba[idx+1] = byte((1.0 - t) * 255)
			rgba[idx+2] = 128
			rgba[idx+3] = 255
		}
	}
	return rgba
}

func verticalGradient(w, h int) []byte {
	rgba := make([]byte, w*h*4)
	for y := 0; y < h; y++ {
		t := float64(y) / math.Max(float64(h-1), 1)
		for x := 0; x < w; x++ {
			idx := (y*w + x) * 4
			rgba[idx] = byte(t * 255)
			rgba[idx+1] = byte(t * 128)
			rgba[idx+2] = byte((1.0 - t) * 255)
			rgba[idx+3] = 255
		}
	}
	return rgba
}

func gamutFromString(s string) Gamut {
	switch s {
	case "sRGB":
		return GamutSRGB
	case "Display P3":
		return GamutDisplayP3
	case "Adobe RGB":
		return GamutAdobeRGB
	case "BT.2020":
		return GamutBT2020
	case "ProPhoto RGB":
		return GamutProPhotoRGB
	default:
		// Falling back to sRGB turned an unrecognised gamut into a hash
		// mismatch on an unrelated assertion. Fail where the cause is, as the
		// Rust and C vector harnesses do.
		panic(fmt.Sprintf("unknown gamut in spec vector: %q", s))
	}
}

// ── integration-encode.json ──────────────────────────────────────────────────

type encodeTestCase struct {
	Name  string `json:"name"`
	Input struct {
		Width  int    `json:"width"`
		Height int    `json:"height"`
		Gamut  string `json:"gamut"`
		Tier   int    `json:"tier"`
		RGBA   []int  `json:"rgba"`
	} `json:"input"`
	Expected struct {
		Hash         []int  `json:"hash"`
		AverageColor [4]int `json:"average_color"`
	} `json:"expected"`
}

func TestIntegrationEncode(t *testing.T) {
	data, err := os.ReadFile("../spec/test-vectors/integration-encode.json")
	if err != nil {
		t.Fatalf("integration-encode.json not found: %v", err)
	}
	var cases []encodeTestCase
	if err := json.Unmarshal(data, &cases); err != nil {
		t.Fatalf("parse integration-encode.json: %v", err)
	}
	if len(cases) == 0 {
		t.Fatal("no encode vectors found")
	}
	for _, tc := range cases {
		t.Run(tc.Name, func(t *testing.T) {
			rgba := make([]byte, len(tc.Input.RGBA))
			for i, v := range tc.Input.RGBA {
				rgba[i] = byte(v)
			}
			ch := EncodeWithQuality(tc.Input.Width, tc.Input.Height, rgba, gamutFromString(tc.Input.Gamut), uint8(tc.Input.Tier))
			if len(ch.Hash) != len(tc.Expected.Hash) {
				t.Fatalf("hash length = %d, want %d", len(ch.Hash), len(tc.Expected.Hash))
			}
			for i, want := range tc.Expected.Hash {
				if int(ch.Hash[i]) != want {
					t.Errorf("hash[%d] = %d, want %d", i, ch.Hash[i], want)
				}
			}
			r, g, b, a := ch.AverageColor()
			avg := [4]int{int(r), int(g), int(b), int(a)}
			for i := 0; i < 4; i++ {
				if avg[i] != tc.Expected.AverageColor[i] {
					t.Errorf("average_color[%d] = %d, want %d", i, avg[i], tc.Expected.AverageColor[i])
				}
			}
		})
	}
}

// ── integration-decode.json ──────────────────────────────────────────────────

type decodeTestCase struct {
	Name  string `json:"name"`
	Input struct {
		Hash []int `json:"hash"`
	} `json:"input"`
	Expected struct {
		Width  int   `json:"width"`
		Height int   `json:"height"`
		RGBA   []int `json:"rgba"`
	} `json:"expected"`
}

func TestIntegrationDecode(t *testing.T) {
	data, err := os.ReadFile("../spec/test-vectors/integration-decode.json")
	if err != nil {
		t.Fatalf("integration-decode.json not found: %v", err)
	}
	var cases []decodeTestCase
	if err := json.Unmarshal(data, &cases); err != nil {
		t.Fatalf("parse integration-decode.json: %v", err)
	}
	if len(cases) == 0 {
		t.Fatal("no decode vectors found")
	}
	for _, tc := range cases {
		t.Run(tc.Name, func(t *testing.T) {
			hashBytes := make([]byte, len(tc.Input.Hash))
			for i, v := range tc.Input.Hash {
				hashBytes[i] = byte(v)
			}
			ch, err := FromBytes(hashBytes)
			if err != nil {
				t.Fatalf("FromBytes rejected a spec vector: %v", err)
			}
			w, h, rgba := ch.Decode()
			if w != tc.Expected.Width {
				t.Errorf("width = %d, want %d", w, tc.Expected.Width)
			}
			if h != tc.Expected.Height {
				t.Errorf("height = %d, want %d", h, tc.Expected.Height)
			}
			for i, want := range tc.Expected.RGBA {
				if int(rgba[i]) != want {
					t.Errorf("rgba[%d] = %d, want %d", i, rgba[i], want)
				}
			}
		})
	}
}

// ── integration-decode-capped.json ────────────────────────────────────────────

type cappedDecodeTestCase struct {
	Name  string `json:"name"`
	Input struct {
		Hash     []int `json:"hash"`
		MaxWidth int   `json:"max_width"`
		MaxHt    int   `json:"max_height"`
	} `json:"input"`
	Expected struct {
		Width  int   `json:"width"`
		Height int   `json:"height"`
		RGBA   []int `json:"rgba"`
	} `json:"expected"`
}

func TestIntegrationDecodeCapped(t *testing.T) {
	data, err := os.ReadFile("../spec/test-vectors/integration-decode-capped.json")
	if err != nil {
		t.Fatalf("integration-decode-capped.json not found: %v", err)
	}
	var cases []cappedDecodeTestCase
	if err := json.Unmarshal(data, &cases); err != nil {
		t.Fatalf("parse integration-decode-capped.json: %v", err)
	}
	if len(cases) == 0 {
		t.Fatal("no capped decode vectors found")
	}
	for _, tc := range cases {
		t.Run(tc.Name, func(t *testing.T) {
			hashBytes := make([]byte, len(tc.Input.Hash))
			for i, v := range tc.Input.Hash {
				hashBytes[i] = byte(v)
			}
			ch, err := FromBytes(hashBytes)
			if err != nil {
				t.Fatalf("FromBytes rejected a spec vector: %v", err)
			}
			w, h, rgba := ch.DecodeCapped(tc.Input.MaxWidth, tc.Input.MaxHt)
			if w != tc.Expected.Width {
				t.Errorf("width = %d, want %d", w, tc.Expected.Width)
			}
			if h != tc.Expected.Height {
				t.Errorf("height = %d, want %d", h, tc.Expected.Height)
			}
			for i, want := range tc.Expected.RGBA {
				if int(rgba[i]) != want {
					t.Errorf("rgba[%d] = %d, want %d", i, rgba[i], want)
				}
			}
		})
	}
}

// ── property tests ─────────────────────────────────────────────────────────────

// tierByteLengths is the opaque-image hash length for each tier code, indexed
// by the code — the table spec §3.3 tabulates.
var tierByteLengths = [MaxTier + 1]int{21, 32, 108, 411, 1623}

// The byte length is a function of the tier alone, so assert all five rather
// than only the default.
func TestEachTierEncodesToItsDocumentedLength(t *testing.T) {
	rgba := solidImage(4, 4, 128, 128, 128, 255)
	for tier, want := range tierByteLengths {
		ch := EncodeWithQuality(4, 4, rgba, GamutSRGB, uint8(tier))
		if len(ch.Hash) != want {
			t.Errorf("tier %d: hash length = %d, want %d", tier, len(ch.Hash), want)
		}
	}
	if got := len(Encode(4, 4, rgba, GamutSRGB).Hash); got != tierByteLengths[DefaultTier] {
		t.Errorf("Encode default length = %d, want %d", got, tierByteLengths[DefaultTier])
	}
}

// The tier codes are declared here as Go consts for idiom, but the format
// defines them and the C ABI exports them. Assert the two agree, so a renumber
// in the core cannot leave this package quietly one code behind.
func TestTierConstantsMatchTheCABI(t *testing.T) {
	abiCompact, abiDefault, abiMax := abiTierCodes()
	for _, tc := range []struct {
		name    string
		go_, c_ uint8
	}{
		{"CompactTier", CompactTier, abiCompact},
		{"DefaultTier", DefaultTier, abiDefault},
		{"MaxTier", MaxTier, abiMax},
	} {
		if tc.go_ != tc.c_ {
			t.Errorf("%s = %d in Go, %d in the C ABI", tc.name, tc.go_, tc.c_)
		}
	}
}

// The header is self-describing, so a length that disagrees with it is rejected
// at construction — not deferred to Decode.
func TestFromBytesRejectsWrongLength(t *testing.T) {
	valid := Encode(4, 4, solidImage(4, 4, 128, 64, 32, 255), GamutSRGB).Hash

	for _, tc := range []struct {
		name  string
		bytes []byte
	}{
		{"one byte short", valid[:len(valid)-1]},
		{"one byte long", append(append([]byte{}, valid...), 0)},
		{"empty", []byte{}},
		{"reserved tier code", append([]byte{(MaxTier + 1) << 3}, valid[1:]...)},
		{"reserved bit set", append([]byte{valid[0] | 0x80}, valid[1:]...)},
	} {
		t.Run(tc.name, func(t *testing.T) {
			if _, err := FromBytes(tc.bytes); !errors.Is(err, ErrInvalidHash) {
				t.Errorf("FromBytes accepted %d bytes (err = %v)", len(tc.bytes), err)
			}
		})
	}
}

func TestFromBytesRoundtrip(t *testing.T) {
	ch := Encode(4, 4, solidImage(4, 4, 128, 64, 32, 255), GamutSRGB)
	got, err := FromBytes(ch.Hash)
	if err != nil {
		t.Fatalf("FromBytes rejected its own output: %v", err)
	}
	if !bytes.Equal(got.Hash, ch.Hash) {
		t.Error("FromBytes roundtrip failed")
	}
}

// FromBytes copies, so a later write to the caller's slice cannot invalidate a
// ChromaHash that already validated.
func TestFromBytesCopiesTheInput(t *testing.T) {
	src := Encode(4, 4, solidImage(4, 4, 128, 64, 32, 255), GamutSRGB).Hash
	ch, err := FromBytes(src)
	if err != nil {
		t.Fatalf("FromBytes: %v", err)
	}
	src[0] ^= 0xff
	if ch.Hash[0] == src[0] {
		t.Error("FromBytes aliased the caller's slice")
	}
}

// Decoded dimensions come from the aspect byte and the tier's raster. A range
// check wide enough to pass for every tier cannot tell them apart, so assert
// the values.
func TestDecodedDimensionsFollowTheTierRaster(t *testing.T) {
	rgba := solidImage(4, 4, 128, 64, 32, 255)
	for tier, edge := range map[uint8]int{0: 32, 1: 32, 2: 64, 3: 128, 4: 256} {
		w, h, pixels := EncodeWithQuality(4, 4, rgba, GamutSRGB, tier).Decode()
		if w != edge || h != edge {
			t.Errorf("tier %d: decoded %dx%d, want %dx%d", tier, w, h, edge, edge)
		}
		if len(pixels) != w*h*4 {
			t.Errorf("tier %d: pixel length %d, want %d", tier, len(pixels), w*h*4)
		}
	}
}

File: bindings/uniffi/jvm/src/test/kotlin/io/chromahash/jvm/ChromaHashTest.kt
package io.chromahash.jvm

import io.chromahash.ffi.BatchEncoder
import io.chromahash.ffi.ChromaHash
import io.chromahash.ffi.ChromaHashException
import io.chromahash.ffi.Gamut
import io.chromahash.ffi.ImageInput
import io.chromahash.ffi.compactTier
import io.chromahash.ffi.defaultTier
import io.chromahash.ffi.formatVersion
import io.chromahash.ffi.maxTier
import org.json.JSONArray
import java.io.File
import java.nio.file.Paths
import kotlin.math.abs
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

/**
 * Drives the spec test vectors through the UniFFI binding (`io.chromahash.ffi`),
 * proving the generated Kotlin + JNA + bundled native lib produce byte-identical
 * output to the Rust reference. This is the JVM module's enforced correctness gate.
 */
class ChromaHashTest {
    private fun specVectorsDir(): File {
        // <repo>/bindings/uniffi/jvm  ->  <repo>/spec/test-vectors
        val cwd = Paths.get("").toAbsolutePath().toFile()
        return cwd.resolve("../../../spec/test-vectors")
    }

    /**
     * A missing or empty vector file is a broken gate, not a reason to pass.
     * This used to return null on a missing file and every caller did `?: return`,
     * so renaming a vector file turned the enforced correctness gate into four
     * tests that asserted nothing and reported green.
     */
    private fun loadJsonArray(name: String): JSONArray {
        val f = specVectorsDir().resolve(name)
        assertTrue(f.exists(), "spec vector file not found: ${f.absolutePath}")
        val arr = JSONArray(f.readText())
        assertTrue(arr.length() > 0, "spec vector file is empty: $name")
        return arr
    }

    private fun gamutFromName(name: String): Gamut =
        when (name) {
            "sRGB" -> Gamut.SRGB
            "Display P3" -> Gamut.DISPLAY_P3
            "Adobe RGB" -> Gamut.ADOBE_RGB
            "BT.2020" -> Gamut.BT2020
            "ProPhoto RGB" -> Gamut.PRO_PHOTO_RGB
            // Falling back to sRGB turned an unrecognised gamut into a hash
            // mismatch on an unrelated assertion. Fail where the cause is.
            else -> throw IllegalArgumentException("unknown gamut in spec vector: $name")
        }

    @Test
    fun `spec vectors directory exists`() {
        val dir = specVectorsDir()
        assertTrue(dir.exists(), "spec vectors dir not found at ${dir.absolutePath}")
    }

    @Test
    fun `integration encode spec vectors`() {
        val arr = loadJsonArray("integration-encode.json")
        for (i in 0 until arr.length()) {
            val tc = arr.getJSONObject(i)
            val name = tc.getString("name")
            val input = tc.getJSONObject("input")
            val width = input.getInt("width")
            val height = input.getInt("height")
            val gamut = gamutFromName(input.getString("gamut"))
            val tier = input.getInt("tier")
            val rgbaArr = input.getJSONArray("rgba")
            val rgba = ByteArray(rgbaArr.length()) { (rgbaArr.getInt(it) and 0xFF).toByte() }
            val expected = tc.getJSONObject("expected")
            val expectedHashArr = expected.getJSONArray("hash")
            val expectedHash =
                ByteArray(expectedHashArr.length()) { (expectedHashArr.getInt(it) and 0xFF).toByte() }

            ChromaHash.encodeWithQuality(width.toUInt(), height.toUInt(), rgba, gamut, tier.toUByte()).use { hash ->
                assertContentEquals(expectedHash, hash.asBytes(), "$name: hash mismatch")
                if (expected.has("average_color")) {
                    val avgArr = expected.getJSONArray("average_color")
                    val avg = hash.averageColor()
                    assertEquals(avgArr.getInt(0), avg.r, "$name: avg.r")
                    assertEquals(avgArr.getInt(1), avg.g, "$name: avg.g")
                    assertEquals(avgArr.getInt(2), avg.b, "$name: avg.b")
                    assertEquals(avgArr.getInt(3), avg.a, "$name: avg.a")
                }
            }
        }
    }

    @Test
    fun `integration decode spec vectors`() {
        val arr = loadJsonArray("integration-decode.json")
        for (i in 0 until arr.length()) {
            val tc = arr.getJSONObject(i)
            val name = tc.getString("name")
            val input = tc.getJSONObject("input")
            val hashArr = input.getJSONArray("hash")
            val hashBytes = ByteArray(hashArr.length()) { (hashArr.getInt(it) and 0xFF).toByte() }
            val expected = tc.getJSONObject("expected")
            val expectedW = expected.getInt("width")
            val expectedH = expected.getInt("height")
            val expectedRgba = expected.getJSONArray("rgba")

            ChromaHash.fromBytes(hashBytes).use { hash ->
                val result = hash.decode()
                assertEquals(expectedW, result.width, "$name: width")
                assertEquals(expectedH, result.height, "$name: height")
                assertEquals(expectedRgba.length(), result.rgba.size, "$name: rgba length")
                for (j in 0 until result.rgba.size) {
                    val got = result.rgba[j].toInt() and 0xFF
                    assertTrue(
                        abs(got - expectedRgba.getInt(j)) <= 1,
                        "$name: rgba[$j] expected ${expectedRgba.getInt(j)}, got $got",
                    )
                }
            }
        }
    }

    @Test
    fun `integration decode capped spec vectors`() {
        val arr = loadJsonArray("integration-decode-capped.json")
        for (i in 0 until arr.length()) {
            val tc = arr.getJSONObject(i)
            val name = tc.getString("name")
            val input = tc.getJSONObject("input")
            val hashArr = input.getJSONArray("hash")
            val hashBytes = ByteArray(hashArr.length()) { (hashArr.getInt(it) and 0xFF).toByte() }
            val maxW = input.getInt("max_width")
            val maxH = input.getInt("max_height")
            val expected = tc.getJSONObject("expected")
            val expectedW = expected.getInt("width")
            val expectedH = expected.getInt("height")
            val expectedRgba = expected.getJSONArray("rgba")

            ChromaHash.fromBytes(hashBytes).use { hash ->
                val result = hash.decodeCapped(maxW.toUInt(), maxH.toUInt())
                assertEquals(expectedW, result.width, "$name: width")
                assertEquals(expectedH, result.height, "$name: height")
                assertEquals(expectedRgba.length(), result.rgba.size, "$name: rgba length")
                for (j in 0 until result.rgba.size) {
                    val got = result.rgba[j].toInt() and 0xFF
                    assertTrue(
                        abs(got - expectedRgba.getInt(j)) <= 1,
                        "$name: rgba[$j] expected ${expectedRgba.getInt(j)}, got $got",
                    )
                }
            }
        }
    }

    private fun solid4x4(): ByteArray =
        ByteArray(64) { i ->
            when (i % 4) {
                0 -> 128.toByte()
                1 -> 64.toByte()
                2 -> 32.toByte()
                else -> 255.toByte()
            }
        }

    /**
     * A fixed length is not a valid assertion about any tier: each has its own,
     * and a bare 16/64 pair says nothing about the ones in between. Derive each
     * length from a real encode, then bracket it.
     */
    @Test
    fun `fromBytes accepts every tier and rejects wrong lengths`() {
        val rgba = solid4x4()
        for (tier in 0..MAX_TIER) {
            val encoded =
                ChromaHash.encodeWithQuality(4u, 4u, rgba, Gamut.SRGB, tier.toUByte()).use {
                    it.asBytes()
                }
            ChromaHash.fromBytes(encoded).use { /* accepted */ }

            assertFailsWith<Exception>("tier $tier: accepted a buffer one byte short") {
                ChromaHash.fromBytes(encoded.copyOf(encoded.size - 1))
            }
            assertFailsWith<Exception>("tier $tier: accepted a buffer one byte long") {
                ChromaHash.fromBytes(encoded.copyOf(encoded.size + 1))
            }
        }
    }

    /**
     * The byte length is a function of the tier alone, so it is checkable
     * exactly rather than as a range. These are the lengths spec §3.3 tabulates.
     */
    @Test
    fun `each tier encodes to its documented length`() {
        val rgba = solid4x4()
        val expected = mapOf(0 to 21, 1 to 32, 2 to 108, 3 to 411, 4 to 1623)
        for ((tier, bytes) in expected) {
            ChromaHash.encodeWithQuality(4u, 4u, rgba, Gamut.SRGB, tier.toUByte()).use { hash ->
                assertEquals(bytes, hash.asBytes().size, "tier $tier byte length")
            }
        }
    }

    /**
     * Decoded dimensions come from the aspect byte and the tier's raster, not
     * from the input size — so assert the values, not a range wide enough to
     * pass for any of them.
     */
    @Test
    fun `decoded dimensions follow the tier raster`() {
        val rgba = solid4x4()
        val expected = mapOf(0 to 32, 1 to 32, 2 to 64, 3 to 128, 4 to 256)
        for ((tier, edge) in expected) {
            ChromaHash.encodeWithQuality(4u, 4u, rgba, Gamut.SRGB, tier.toUByte()).use { hash ->
                val result = hash.decode()
                assertEquals(edge, result.width.toInt(), "tier $tier width")
                assertEquals(edge, result.height.toInt(), "tier $tier height")
                assertEquals(result.width.toInt() * result.height.toInt() * 4, result.rgba.size)
            }
        }
    }

    /**
     * The core panics on invalid input, and a panic across the FFI boundary is
     * undefined behaviour — so the binding validates first and throws a typed
     * error, matching the C ABI's status codes.
     */
    @Test
    fun `encode rejects invalid dimensions`() {
        assertFailsWith<ChromaHashException.InvalidDimensions> {
            ChromaHash.encode(0u, 4u, ByteArray(0), Gamut.SRGB)
        }
        assertFailsWith<ChromaHashException.InvalidDimensions> {
            ChromaHash.encode(4u, 0u, ByteArray(0), Gamut.SRGB)
        }
        // rgba shorter than w * h * 4
        assertFailsWith<ChromaHashException.InvalidLength> {
            ChromaHash.encode(4u, 4u, ByteArray(63), Gamut.SRGB)
        }
    }

    @Test
    fun `encodeWithQuality rejects a reserved tier code`() {
        val rgba = solid4x4()
        assertFailsWith<ChromaHashException.InvalidTier> {
            ChromaHash.encodeWithQuality(4u, 4u, rgba, Gamut.SRGB, (MAX_TIER + 1).toUByte())
        }
    }

    /**
     * Pins the tier down to the byte count. Comparing the batch against the
     * serial path alone would pass if both silently used one tier.
     */
    @Test
    fun `batch encoding honors each item's tier`() {
        val rgba = solid4x4()
        val items = (0..MAX_TIER).map { ImageInput(4u, 4u, rgba, Gamut.SRGB, it.toUByte()) }
        BatchEncoder().use { encoder ->
            val lengths = encoder.encodeBatch(items).map { it.use { h -> h.asBytes().size } }
            assertContentEquals(listOf(21, 32, 108, 411, 1623), lengths)
        }
    }

    /**
     * An item with no explicit tier must match `encode` — the codes are ordered
     * by quality, so a zero default would be the 21-byte compact tier.
     */
    @Test
    fun `batch encoding defaults to the default tier`() {
        val rgba = solid4x4()
        val expected = ChromaHash.encode(4u, 4u, rgba, Gamut.SRGB).use { it.asBytes() }
        BatchEncoder().use { encoder ->
            val batched =
                encoder.encodeBatch(listOf(ImageInput(4u, 4u, rgba, Gamut.SRGB))).single().use {
                    it.asBytes()
                }
            assertContentEquals(expected, batched)
        }
    }

    @Test
    fun `batch encoding rejects a reserved tier code, naming the item`() {
        val rgba = solid4x4()
        val items =
            listOf(
                ImageInput(4u, 4u, rgba, Gamut.SRGB),
                ImageInput(4u, 4u, rgba, Gamut.SRGB, (MAX_TIER + 1).toUByte()),
            )
        BatchEncoder().use { encoder ->
            val e =
                assertFailsWith<ChromaHashException.InvalidTier> { encoder.encodeBatch(items) }
            assertTrue(e.message!!.contains("item 1"), "error should name the item: ${e.message}")
        }
    }

    /**
     * The tier codes reach Kotlin through the FFI rather than being restated
     * here, so this asserts the ordering the format guarantees.
     */
    @Test
    fun `tier constants come from the core`() {
        assertEquals(0, compactTier().toInt())
        assertEquals(MAX_TIER, maxTier().toInt())
        assertTrue(compactTier() < defaultTier() && defaultTier() < maxTier())
        assertEquals(0, formatVersion().toInt())
    }

    private companion object {
        /** Highest tier code the format defines; codes above it are reserved. */
        val MAX_TIER = maxTier().toInt()
    }
}
